flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/schema/TableSourceTable.scala

```diff
@@ -19,20 +19,24 @@
 package org.apache.flink.table.plan.schema
 
 import org.apache.calcite.rel.`type`.{RelDataType, RelDataTypeFactory}
+import org.apache.flink.table.{JHashSet, JSet}
 import org.apache.flink.table.calcite.FlinkTypeFactory
 import org.apache.flink.table.plan.stats.FlinkStatistic
 import org.apache.flink.table.sources.{TableSource, TableSourceUtil}
 
+import java.util
+
 /**
   * Abstract class which define the interfaces required to convert a [[TableSource]] to
   * a Calcite Table
   */
 class TableSourceTable[T](
     val tableSource: TableSource[T],
-    val isStreaming: Boolean,
-    val statistic: FlinkStatistic)
```

Comment:
if the `statistic` is deleted, how to store the statistic from catalog?

Reply:
Another scenario: rules (like `PushProjectIntoTableSourceScanRule`) does not change statistics   , so the new TableSource created by the rule could reuse the original TableSource, and avoid to call `TableSource`#getTableStats method which is high cost.
so the ` def copy(statistic: FlinkStatistic): FlinkTable` method defined in `FlinkTable` should not be delete too.

Reply:
1. the statistic from catalog should be restored via `TableSource#getTableStats()`.
2. I think if the TableSource is not changed, then we don't need to re-construct a new `TableSourceTable`. I can reuse the original `TableSourceTable` and avoid calling `TableSource#getTableStats` again.

Reply:
1. `TableSource` should be decoupled with `Catalog`, or add a method like `setTableStats` for `TableSource` interface.
2. The TableSource may be changed,  e.g. project push down into table source

Reply:
1. Yes, maybe we need something like `setTableStats`, but this is out of the scope of this issue.
2. If the TableSource is changed, shouldn't we always to create a new `TableSourceTable` and `getTableStats()` again? How do we know the stats is not changed?

Reply:
1. I do not think it‘s a good idea to add `setTableStats ` for `TableSource`, one big reason is each `TableSource` must implement this method and default implementation does not work due to a TableSource does not know whether the catalog could provide its stats.

3. yes, `TableSource` is immutable in `TableSourceTable`. currently, whether use the original stats or use unknown stats for new TableSourceTable is decided in each rule. you could reference this related code in `PushProjectIntoTableSourceScanRule` and `PushFilterIntoTableSourceScanRule` in blink inner code.